server/src/copilot/utils/lock.ts

```diff
@@ -0,0 +1,31 @@
+export class Lock {
+    private isLocked: boolean;
+    private waiting: Array<() => void>;
+
+    constructor() {
+        this.isLocked = false;
+        this.waiting = [];
+    }
+
+    acquire(): Promise<void> {
+        return new Promise((resolve) => {
+            if (!this.isLocked) {
+                this.isLocked = true;
+                resolve();
+            } else {
+                this.waiting.push(resolve);
```

Comment:
Seems like we should reject is we have too many folks waiting to acquire a lock...?

Reply:
Fixed, same as above, now we terminate requests if queue is full.